src/Gaffer/Animation.cpp

```diff
@@ -549,6 +549,112 @@ Animation::KeyPtr Animation::CurvePlug::addKey( const Animation::KeyPtr &key, co
 	return clashingKey;
 }
 
+Animation::KeyPtr Animation::CurvePlug::insertKey( const float time )
+{
+	return insertKeyInternal( time, nullptr );
+}
+
+Animation::KeyPtr Animation::CurvePlug::insertKey( const float time, const float value )
+{
+	return insertKeyInternal( time, & value );
+}
+
+Animation::KeyPtr Animation::CurvePlug::insertKeyInternal( const float time, const float* const value )
+{
+	// find span keys for time
+
+	KeyPtr key;
+	Key* lo = nullptr;
+	Key* hi = nullptr;
+
+	Keys::iterator hiIt = m_keys.lower_bound( time );
+	if( hiIt != m_keys.end() )
+	{
+		hi = &( *( hiIt ) );
+
+		if( hi->m_time == time )
+		{
+			key = hi;
+		}
+		else if( hiIt != m_keys.begin() )
+		{
+			lo = &( *( --hiIt ) );
+
+			if( lo->m_time == time )
+			{
+				key = lo;
+			}
+		}
+	}
+	else
+	{
+		lo = finalKey();
+	}
+
+	// if key already exists at time then return it with updated value, otherwise if time is
+	// outside existing range of keys, there is no way currently to extrapolate a value so
+	// if no value has been provided then return KeyPtr().
+
+	if( key )
+	{
+		if( value != nullptr )
+		{
+			key->setValue( *value );
+		}
+
+		return key;
+	}
+	else if( !( lo && hi ) && ( value == nullptr ) )
+	{
+		return key;
+	}
+
+	// get interpolation
+
+	Interpolation interpolation = Gaffer::Animation::defaultInterpolation();
+
+	if( lo && hi )
+	{
+		interpolation = lo->m_interpolation;
+	}
+	else if( const Key* const kf = firstKey() )
+	{
+		interpolation = kf->m_interpolation;
+	}
+
+	// create key
+
+	key.reset( new Animation::Key( time, ( ( value != nullptr ) ? ( *value ) : 0.f ), interpolation ) );
+
+	// if specified value is the same as the evaluated value of the curve then bisect span.
+
+	if( ( lo && hi ) && ( ( value == nullptr ) || ( evaluate( time ) == ( *value ) ) ) )
```

Comment:
I don't think I understand the `|| ( evaluate( time ) == ( *value ) )` bit here. If that's true, then we already have the value we want, so don't need to do anything at all. Perhaps it's because you want to perform bisection for this case when the full bisection code is added? I don't think that would desirable - the semantics are clearer if `insert( time )` alway does bisection and `insert( time, value )` never does, and float precision doesn't come into it. That's while we don't have the auto-slope path fully implemented - when we do implement that I would expect it to be equivalent to bisection when `value` is on the curve, so that there's a smooth transition between the two anyway.